src/Logary/Registry.fs

```diff
@@ -201,51 +198,94 @@ module Advanced =
       sb.ToString()
 
-    let pollMetrics registry =
-      registry <-- PollMetrics
+    let pollMetrics (registry: Ch<RegistryMessage>) =
+      Job.start (Ch.send registry PollMetrics)
 
     /// The state for the Registry
     type RegistryState =
       { /// the supervisor, not of the registry actor, but of the targets and probes and health checks.
-        supervisor     : IActor
-        schedules      : (string * CancellationTokenSource) list
-        pollMetrics    : CancellationTokenSource option }
+        supervisor  : Supervisor.Instance
+        schedules   : (PointName * Cancellation) list
+        pollMetrics : Cancellation option }
 
     /// The registry function that works as an actor.
-    let rec registry conf sup sched (inbox : IActor<_>) =
-      let targets = conf.targets |> LogaryConfLenses.targetActors_.get
-      let lgr = conf.metadata.logger
-      let regPath = "Logary.Registry.registry"
-      let log = LogLine.setPath regPath >> Logger.log lgr
-
-      let rec init state = async {
-        let ctss = System.Collections.Generic.List<_>()
-        // init sampling of metrics
-        for m in conf.metrics do
-          let (conf, actor) = m.Value
-          match actor with
-          | Some actor ->
-            let! cts = Scheduling.schedule sched Metric.sample actor conf.sampling (Some conf.sampling)
-            ctss.Add( m.Key, cts )
-          | None -> ()
+    let rec registry conf sup sched (inbox : Ch<RegistryMessage>) =
+      let targets = conf.targets |> Seq.map (fun kv -> kv.Value |> snd |> Option.get)
+      let regPath = PointName ["Logary"; "Registry"; "registry" ]
+      let log = Message.setName regPath >> Logger.log conf.runtimeInfo.logger
+
+      /// In the shutdown state, the registry doesn't respond to messages, but rather tries to
+      /// flush and shut down all targets, doing internal logging as it goes.
+      let shutdown state (ackCh : Ch<unit>) (nack : Promise<unit>) = job {
+        do! Message.info "shutdown metrics polling" |> log
+        do! defaultArg (Option.map Cancellation.cancel state.pollMetrics) (Job.unit ())
+
+        do! Message.info "cancel schedules" |> log
+        for (_, x) in state.schedules do
+          do! Cancellation.cancel x
+
+        do! Message.info "shutdown targets" |> log
+        let! allShutdown =
+          targets
+          |> Seq.pjmap (fun t -> Target.shutdown t  ^->. Success Ack
+                                 <|> timeOutMillis 200 ^->. TimedOut)
+
+        let stopped, failed =
+          allShutdown
+          |> Seq.groupBy wasSuccessful
+          |> bisectSuccesses
+
+        if List.isEmpty failed then
+          do! Message.info "shutdown Ack" |> log
+        else
+          let msg = sprintf "failed target shutdowns:%s%s" nl (toTextList failed)
+          do! Message.errorf "shutdown Nack%s%s" nl msg
+              |> Message.field "failed" (Value.fromObject failed)
+              |> log
+
+        do! Ch.give ackCh () <|> nack
```

Comment:
Note that here the semantics don't quite work out, because even if the nack is set (client no longer waits) the shutdown isn't cancelled.  Use nack when the server can (and should) cancel the operation when client stops waiting.

One way to treat the shutdown is as an idempotent (and non cancellable) operation.  In that case you would use `prepare` to async send (`*<+`) an `IVar` to the server that the server then fills when shutdown is complete.

More complex protocols are also possible.  In this case it would likely be possible to make a kind of two-phase shutdown protocol.  Things that need to be shutdown could be in three states: "active" (accepting all messages and acting upon them), "shutting down" (not accepting messages, but still having resources intact) and "disposed" (not accepting messages/responding with error and having resources released).  Upon receiving an asynchronously sent shutdown request, containing a nack, things would go into "shutting down" state by flushing resources and making it trivial to dispose everything.  After everything is "ready to flip the dispose switch", the client would be acknowledged synchronously and everything would be disposed or, in case client has aborted (nack is set and client does not take ack message), everything would be transferred back to active case (meaning that shutdown is cancelled).